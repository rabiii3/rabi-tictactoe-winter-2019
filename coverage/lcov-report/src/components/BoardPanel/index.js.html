<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for src/components/BoardPanel/index.js</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../prettify.css" />
    <link rel="stylesheet" href="../../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../../index.html">All files</a> / <a href="index.html">src/components/BoardPanel</a> index.js
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">41.67% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>10/24</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">0% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>0/10</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">0% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>0/7</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">52.63% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>10/19</span>
      </div>
    </div>
  </div>
  <div class='status-line low'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91</td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import React from 'react';
import PropTypes from 'prop-types';
import { map, addIndex } from 'ramda';
import { Grid, Typography, withStyles } from '@material-ui/core';
import { GAME_STARTED } from '../../game';
&nbsp;
const styles = {
  container: {
    display: 'flex',
    flexDirection: 'row',
    height: 300,
    width: 300,
    border: '1px solid black',
    borderWidth:'0 1px 1px 0',
    flexWrap: 'wrap',
  },
  square: {
    width: '33.336%',
    height: 100,
    border: '1px solid black',
    borderWidth:' 1px 0 0 1px',
    textAlign: 'center',
    lineHeight: 6.7,
  },
};
&nbsp;
const DeadCell = <span class="fstat-no" title="function not covered" >()</span> =&gt; <span class="cstat-no" title="statement not covered" >&lt;div style={styles.square} /&gt;;</span>
&nbsp;
const PlayedCell = <span class="fstat-no" title="function not covered" >({</span> piece }) =&gt; <span class="cstat-no" title="statement not covered" >&lt;div style={styles.square}&gt;{piece}&lt;/div&gt;;</span>
&nbsp;
PlayedCell.propTypes = {
  piece: PropTypes.string,
};
&nbsp;
const ClickableCell = <span class="fstat-no" title="function not covered" >({</span> onClick }) =&gt; <span class="cstat-no" title="statement not covered" >&lt;div style={styles.square} onClick={onClick} /&gt;;</span>
&nbsp;
ClickableCell.propTypes = {
  onClick: PropTypes.func.isRequired,
};
&nbsp;
const Cell = <span class="fstat-no" title="function not covered" >({</span> currentPlayer, piece, onClick, status }) =&gt; {
<span class="cstat-no" title="statement not covered" >  if (status === GAME_STARTED) {</span>
<span class="cstat-no" title="statement not covered" >    if (!piece) {</span>
<span class="cstat-no" title="statement not covered" >      if (currentPlayer.isComputer) <span class="cstat-no" title="statement not covered" >return &lt;DeadCell /&gt;;</span></span>
<span class="cstat-no" title="statement not covered" >      return &lt;ClickableCell onClick={onClick} /&gt;;</span>
    }
<span class="cstat-no" title="statement not covered" >    return &lt;PlayedCell piece={piece} /&gt;;</span>
  }
<span class="cstat-no" title="statement not covered" >  return &lt;PlayedCell piece={piece} /&gt;;</span>
};
&nbsp;
Cell.propTypes = {
  currentPlayer: PropTypes.object.isRequired,
  piece: PropTypes.string,
  onClick: PropTypes.func.isRequired,
  status: PropTypes.string.isRequired,
};
&nbsp;
const BoardPanel = <span class="fstat-no" title="function not covered" >({</span> board, currentPlayer, status, onClick, classes }) =&gt; {
  const mapIndex = <span class="cstat-no" title="statement not covered" >addIndex(map);</span>
<span class="cstat-no" title="statement not covered" >  return (</span>
    &lt;Grid container direction="column" justify="space-around" spacing={16}&gt;
      &lt;Grid item&gt;
        &lt;Typography variant="h6" gutterBottom&gt;
          {status === GAME_STARTED &amp;&amp; currentPlayer ? 'Current Player: ' + currentPlayer.name : 'GAME OVER'}
        &lt;/Typography&gt;
      &lt;/Grid&gt;
      &lt;Grid item&gt;
        &lt;div className={classes.container}&gt;
          {mapIndex(
<span class="fstat-no" title="function not covered" >            (c</span>ell, i) =&gt; (
<span class="cstat-no" title="statement not covered" >              &lt;Cell key={i} currentPlayer={currentPlayer} piece={cell} status={status} onClick={<span class="fstat-no" title="function not covered" >()</span> =&gt; <span class="cstat-no" title="statement not covered" >onClick(i)}</span> /&gt;</span>
            ),
            board,
          )}
        &lt;/div&gt;
      &lt;/Grid&gt;
    &lt;/Grid&gt;
  );
};
&nbsp;
BoardPanel.propTypes = {
  board: PropTypes.array.isRequired,
  currentPlayer: PropTypes.object,
  status: PropTypes.string.isRequired,
  classes: PropTypes.object.isRequired,
  onClick: PropTypes.func.isRequired,
};
&nbsp;
export default withStyles(styles)(BoardPanel);
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Thu Mar 14 2019 06:36:09 GMT+0100 (Central European Standard Time)
</div>
</div>
<script src="../../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../../sorter.js"></script>
</body>
</html>
