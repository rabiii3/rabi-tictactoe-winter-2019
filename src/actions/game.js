import { GAME_STARTED, getNewBoard, getNextPlayer, updateBoard, GAME_OVER, isBoardFull, randomPlayer } from '../game';
import { requestComputerPlay, hasAWinner } from '../request';
import { loadFruits } from './fruit';

export const STARTED = 'STARTED';
export const PLAYED = 'PLAYED';
export const PLAYER_CHANGED = 'PLAYER_CHANGED';
export const ENDED = 'ENDED';

export const startGameAction = () => (dispatch, getState) => {
  const {
    game: { player, computer },
  } = getState();
  dispatch({
    type: STARTED,
    payloads: {
      board: getNewBoard(),
      status: GAME_STARTED,
      currentPlayer: randomPlayer(player, computer),
    },
  });
  const {
    game: { currentPlayer },
  } = getState();
  if (currentPlayer.isComputer) dispatch(computerPlay());
};

export const playerPlay = index => (dispatch, getState) => {
  const {
    game: { board, currentPlayer },
  } = getState();
  const newBoard = updateBoard(board, index, currentPlayer);
  dispatch(manageGame(newBoard));
};

const computerPlay = () => (dispatch, getState) => {
  const {
    game: { board, currentPlayer, player, computer },
  } = getState();
  return requestComputerPlay({ board, computer: computer.piece, player: player.piece }).then(({ move }) => {
    const newBoard = updateBoard(board, move, currentPlayer);
    setTimeout(() => dispatch(manageGame(newBoard)), 1000);
  });
};

export const manageGame = newBoard => (dispatch, getState) => {
  dispatch({
    type: PLAYED,
    payloads: {
      board: newBoard,
    },
  });

  const {
    game: { board, currentPlayer, player, computer },
  } = getState();
  hasAWinner({ board }).then(({ winner }) => {
    if (winner || isBoardFull(board)) {
      dispatch(loadFruits()).then(() =>
        dispatch({
          type: ENDED,
          payloads: {
            result: winner,
            player,
            computer,
            status: GAME_OVER,
          },
        }),
      );
    } else {
      dispatch({
        type: PLAYER_CHANGED,
        payloads: {
          currentPlayer: getNextPlayer(currentPlayer, player, computer),
        },
      });

      const {
        game: { currentPlayer: cp },
      } = getState();
      console.log('======currennt player :', cp);
      if (cp.isComputer) dispatch(computerPlay());
    }
  });
};
