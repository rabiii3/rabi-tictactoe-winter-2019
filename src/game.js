import { indexOf, update, propOr } from 'ramda';

export const X = 'x';

export const O = 'o';

export const GAME_OVER = 'GAME_OVER';

export const GAME_STARTED = 'GAME_STARTED';

export const isBoardFull = board => indexOf(null, board) < 0;

export const getStatus = board => (isBoardFull(board) ? GAME_OVER : GAME_STARTED);

export const getNewComputerBoard = (board, computer) => update(indexOf(null, board), computer.piece, board);

export const getNewBoard = () => [null, null, null, null, null, null, null, null, null];

export const getFirstAvailableCell = board => indexOf(null, board);

export const updateBoard = (board, index, currentPlayer) =>
  isBoardFull(board) ? board : update(index, currentPlayer.piece, board);

export const getNextPlayer = (currentPlayer, player, computer) => (currentPlayer.isComputer ? player : computer);

const random = () => Math.round(Math.random());

export const randomPlayer = (player, computer) => {
  const players = [player, computer];
  return players[random()];
};

export const updateFruitBoard = (board, index, icon) => update(index, icon, board);

export const isFruit = piece => (piece != null && propOr(false, 'icon', piece) ? true : false);
