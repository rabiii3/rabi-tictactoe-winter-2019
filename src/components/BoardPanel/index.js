import React from 'react';
import PropTypes from 'prop-types';
import { map, addIndex } from 'ramda';
import { pure } from 'recompose';
import { Grid, Typography, withStyles } from '@material-ui/core';
import { GAME_STARTED, isFruit } from '../../game';

const styles = {
  container: {
    display: 'flex',
    flexDirection: 'row',
    height: 300,
    width: 300,
    border: '1px solid black',
    borderWidth: '0 1px 1px 0',
    flexWrap: 'wrap',
  },
  square: {
    width: '33.336%',
    height: 100,
    border: '1px solid black',
    borderWidth: ' 1px 0 0 1px',
    textAlign: 'center',
    lineHeight: 6.7,
  },
};

const DeadCell = () => <div style={styles.square} />;

const PlayedCell = ({ piece }) => <div style={styles.square}>{piece}</div>;

PlayedCell.propTypes = {
  piece: PropTypes.string,
};

const ClickableCell = ({ onClick }) => <div style={styles.square} onClick={onClick} />;

ClickableCell.propTypes = {
  onClick: PropTypes.func.isRequired,
};

const Fruit = ({ piece }) => {
  const { icon, color } = piece;
  return (
    <div style={styles.square}>
      <i className={`fa fa-${icon}`} style={{ color }} />
    </div>
  );
};

Fruit.propTypes = {
  piece: PropTypes.object,
};

const Cell = ({ currentPlayer, piece, onClick }) => {
  if (isFruit(piece)) return <Fruit piece={piece} />;
  if (!piece) {
    if (currentPlayer.isComputer) return <DeadCell />;
    return <ClickableCell onClick={onClick} />;
  }
  return <PlayedCell piece={piece} />;
};

Cell.propTypes = {
  currentPlayer: PropTypes.object.isRequired,
  piece: PropTypes.string,
  onClick: PropTypes.func.isRequired,
  status: PropTypes.string.isRequired,
};

const PureCell = React.memo(Cell);
//const PureCell = pure(Cell);

const BoardPanel = ({ board, currentPlayer, status, onClick, classes }) => {
  const mapIndex = addIndex(map);
  return (
    <Grid container direction="column" justify="space-around" spacing={16}>
      <Grid item>
        <Typography variant="h6" gutterBottom align="center">
          {status === GAME_STARTED && currentPlayer ? 'Current Player: ' + currentPlayer.name : 'GAME OVER'}
        </Typography>
      </Grid>
      <Grid item>
        <div className={classes.container}>
          {mapIndex(
            (cell, i) => (
              <PureCell key={i} currentPlayer={currentPlayer} piece={cell} status={status} onClick={() => onClick(i)} />
            ),
            board,
          )}
        </div>
      </Grid>
    </Grid>
  );
};

BoardPanel.propTypes = {
  board: PropTypes.array.isRequired,
  currentPlayer: PropTypes.object,
  status: PropTypes.string.isRequired,
  classes: PropTypes.object.isRequired,
  onClick: PropTypes.func.isRequired,
};

export default withStyles(styles)(BoardPanel);
