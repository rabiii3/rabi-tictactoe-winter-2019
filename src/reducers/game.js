import { STARTED, PLAYED, PLAYER_CHANGED, ENDED } from '../actions/game';
import { CELL_FRUIT_LOADED } from '../actions/fruit';
import { updateFruitBoard } from '../game';

export default (state = {}, action) => {
  var i = 0;
  switch (action.type) {
    case STARTED:
      return { ...state, ...action.payloads };
    case PLAYED:
      return { ...state, ...action.payloads };
    case PLAYER_CHANGED:
      return { ...state, ...action.payloads };
    case ENDED: {
      const { status } = action.payloads;
      return { ...state, status };
    }
    case CELL_FRUIT_LOADED: {
      const { board } = state;
      const { index, fruit } = action;
      return { ...state, board: updateFruitBoard(board, index, fruit) };
    }
    default:
      return state;
  }
};
