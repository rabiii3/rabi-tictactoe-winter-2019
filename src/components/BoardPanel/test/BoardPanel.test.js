import React from 'react';
import { shallow } from 'enzyme';
import BoardPanel from '..';
import { X, O, GAME_STARTED } from '../../../game';

describe('BoardPanel', () => {
  it('should render', () => {
    const props = {
      board: [null, X, O],
      currentPlayer: { name, isComputer: false, piece: O },
      onClick: jest.fn,
      status: GAME_STARTED,
    };
    const root = <BoardPanel {...props} />;
    const wrapped = shallow(root);
    expect(wrapped).toMatchSnapshot();
  });
});
