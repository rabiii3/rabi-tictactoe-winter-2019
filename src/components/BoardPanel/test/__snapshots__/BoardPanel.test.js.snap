// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoardPanel should render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WithStyles(BoardPanel)
    board={
      Array [
        null,
        "x",
        "o",
      ]
    }
    currentPlayer={
      Object {
        "isComputer": false,
        "name": "nodejs",
        "piece": "o",
      }
    }
    onClick={[Function]}
    status="GAME_STARTED"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "board": Array [
        null,
        "x",
        "o",
      ],
      "classes": Object {
        "container": "BoardPanel-container-1",
        "square": "BoardPanel-square-2",
      },
      "currentPlayer": Object {
        "isComputer": false,
        "name": "nodejs",
        "piece": "o",
      },
      "onClick": [Function],
      "status": "GAME_STARTED",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "board": Array [
          null,
          "x",
          "o",
        ],
        "classes": Object {
          "container": "BoardPanel-container-1",
          "square": "BoardPanel-square-2",
        },
        "currentPlayer": Object {
          "isComputer": false,
          "name": "nodejs",
          "piece": "o",
        },
        "onClick": [Function],
        "status": "GAME_STARTED",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
